### PR TITLE
Make the IPAM cleanup at the end of the task

### PR DESCRIPTION
**Pull request by richardpen**
### Summary

Fixing #970 

### Implementation details

Previously, the pause container namespace cleanup was invoked in `StopContainer` of task engine. This pr separate the cleanup into two process: clean up the network namespace and release the ip from ipam. Also this pr moves the ipam clean up out of the `StopContainer` to the end of a task life cycle. So that the `ipam del` command is always called when a task moved to stop.

### Testing

- [x] Builds on Linux (`make release`)
- [x] Builds on Windows (`go build -out amazon-ecs-agent.exe ./agent`)
- [x] Unit tests on Linux (`make test`) pass
- [x] Unit tests on Windows (`go test -timeout=25s ./agent/...`) pass
- [x] Integration tests on Linux (`make run-integ-tests`) pass
- [x] Integration tests on Windows (`.\scripts\run-integ-tests.ps1`) pass
- [x] Functional tests on Linux (`make run-functional-tests`) pass
- [x] Functional tests on Windows (`.\scripts\run-functional-tests.ps1`) pass
- [x] Manual test the scenario where the pause container is killed outside agent

New tests cover the changes: 
yes

### Licensing

This contribution is under the terms of the Apache 2.0 License: 
yes

### CODE REVIEW

File: agent/ecscni/plugin.go [hunk 1]
```diff
@@ -135,6 +162,48 @@ func (client *cniClient) constructNetworkConfig(cfg *Config) (*libcni.NetworkCon
...
+		Network: &types.NetConf{
+			Type: ECSBridgePluginName,
+		},
+		Bytes: bridgeConfBytes,
+	}
+
+	return networkConfig, nil
+}
+
+func (client *cniClient) createBridgeConfig(cfg *Config) (*libcni.NetworkConfig, error) {
```
Comment (samuelkarp) [inline agent/ecscni/plugin.go]: It looks like there's a bunch of duplicate code between `createBridgeConfig` and `createBridgeConfigWithIPAM`.  Can we share instead of duplicating?

File: agent/ecscni/plugin.go [hunk 2]
```diff
@@ -135,6 +162,48 @@ func (client *cniClient) constructNetworkConfig(cfg *Config) (*libcni.NetworkCon
...
+		Network: &types.NetConf{
+			Type: ECSBridgePluginName,
+		},
+		Bytes: bridgeConfBytes,
+	}
+
+	return networkConfig, nil
+}
+
+func (client *cniClient) createtENIConfig(cfg *Config) (*libcni.NetworkConfig, error) {
```
Comment (samuelkarp) [inline agent/ecscni/plugin.go]: s/createt/create/

File: agent/ecscni/plugin.go [hunk 3]
```diff
@@ -145,37 +214,70 @@ func (client *cniClient) constructNetworkConfig(cfg *Config) (*libcni.NetworkCon
...
-		Bytes: eniConfBytes,
-	})
+		Bytes: ipamConfigBytes,
+	}
 
-	netconf := &libcni.NetworkConfigList{
+	return networkConfig, nil
+}
+
+// createAddNetworkConfig creates the configuration for invoking plugins
```
Comment (samuelkarp) [inline agent/ecscni/plugin.go]: nit: createNetworkConfig

File: agent/ecscni/plugin.go [hunk 4]
```diff
@@ -145,37 +214,70 @@ func (client *cniClient) constructNetworkConfig(cfg *Config) (*libcni.NetworkCon
...
+func (client *cniClient) createNetworkConfig(cfg *Config, bridgeConfigFunc func(*Config) (*libcni.NetworkConfig, error)) (*libcni.NetworkConfigList, error) {
```
Comment (samuelkarp) [inline agent/ecscni/plugin.go]: It looks like this function is the entrypoint for both `createBridgeConfig` and `createBridgeConfigWithENI`.  Can you move it above those functions?

File: agent/ecscni/plugin.go [hunk 5]
```diff
@@ -145,37 +214,70 @@ func (client *cniClient) constructNetworkConfig(cfg *Config) (*libcni.NetworkCon
...
+		Name:       ECSIPAMPluginName,
+		CNIVersion: client.cniVersion,
+		IPAM:       ipam,
+	}
+
+	ipamConfigBytes, err := json.Marshal(ipamConfig)
 	if err != nil {
-		seelog.Errorf("Marshal eni configuration error: %v", err)
-		return nil, err
+		return nil, errors.Wrapf(err, "Marshal ipam plugin configuration failed")
```
Comment (aaithal) [inline agent/ecscni/plugin.go]: nit: error message should begin with lower case letter. please fix this everywhere.

File: agent/engine/docker_container_engine.go [hunk 6]
```diff
@@ -39,6 +40,7 @@ import (
 
 const (
 	dockerDefaultTag = "latest"
+	ImageNameFormat  = "%s:%s"
```
Comment (aaithal) [inline agent/engine/docker_container_engine.go]: lint: `ImageNameFormat ...` Or better yet, make it local scoped.

File: agent/engine/task_manager.go [hunk 7]
```diff
@@ -323,6 +324,19 @@ func (mtask *managedTask) handleContainerChange(containerChange dockerContainerC
...
+// releaseIPInIPAM releases the ip used by the task for awsvpc
+func (mtask *managedTask) releaseIPInIPAM() {
+	if mtask.ENI == nil {
+		return
+	}
+	seelog.Infof("Releasing ip in IPAM for task used eni, task: %s", mtask.Task.Arn)
+
+	err := mtask.engine.releaseIPInIPAM(mtask.Task)
+	if err != nil {
+		seelog.Infof("Releasing the ip in IPAM failed, task: [%s], err: %v", mtask.Task.Arn, err)
```
Comment (aaithal) [inline agent/engine/task_manager.go]: Why is this `Infof`? And, why is it ok to consume this error?

Comment (richardpen) [reply agent/engine/task_manager.go]: For this kind of error, it's fine because there will be  more available ips than the task, so if one ip was failed to release the next task can still get an available ip. But in rare condition all the release failed, and all the ip has been used then no task can be started, that's why I marked it as `Infof`, any better suggestion here?

Comment (aaithal) [reply agent/engine/task_manager.go]: So, it's probably fine that we consume this error and move on. I'd still like this to be logged as a warning.

**Comment by richardpen** [comment]
@aaithal @samuelkarp I have addressed your comments, please take another review, thanks.

File: agent/engine/docker_container_engine.go [hunk 8]
```diff
@@ -271,14 +274,19 @@ func (dg *dockerGoClient) pullImage(image string, authData *api.RegistryAuthenti
...
-	if image == emptyvolume.Image+":"+emptyvolume.Tag {
+	if image == fmt.Sprintf(imageNameFormat, emptyvolume.Image, emptyvolume.Tag) {
 		scratchErr := dg.createScratchImageIfNotExists()
 		if scratchErr != nil {
 			return CreateEmptyVolumeError{scratchErr}
 		}
 		return nil
 	}
 
+	// Skip pull pause container image, it's already loaded on agent restart
```
Comment (samuelkarp) [inline agent/engine/docker_container_engine.go]: Can you remove this logic for now?  I think I'd rather have this occur at a different layer in the codebase (probably in DockerTaskEngine rather than here, so we can trigger on the container type rather than the image name) and it's orthogonal to the rest of the PR.  I was planning on looking at this on Monday.

File: agent/engine/task_manager.go [hunk 9]
```diff
@@ -151,6 +151,7 @@ func (mtask *managedTask) overseeTask() {
 		llog.Debug("Marking done for this sequence", "seqnum", mtask.StopSequenceNumber)
 		mtask.engine.taskStopGroup.Done(mtask.StopSequenceNumber)
 	}
+	go mtask.releaseIPInIPAM()
```
Comment (samuelkarp) [inline agent/engine/task_manager.go]: Is this idempotent?  You may end up with multiple calls/partial completion if there's an agent restart in the middle.

File: agent/ecscni/plugin.go [hunk 10]
```diff
@@ -84,62 +85,127 @@ func (client *cniClient) SetupNS(cfg *Config) error {
 	return nil
 }
 
+// ReleaseIPResource marks the ip available in the ipam db
+func (client *cniClient) ReleaseIPResource(cfg *Config) error {
```
[the remark this answers is not in the document]
Comment (richardpen) [reply agent/ecscni/plugin.go]: No, just randomly placed, I will fix that.

File: agent/engine/task_manager.go (hunk 9 above)
Comment (richardpen) [reply agent/engine/task_manager.go]: This will ensure the ip is removed if agent restarted, but could emit an error in the ipam logs if  this was called twice before and after restart. One possible solution would be remove the eniattachment from state here and check the eniattachment before calling releasing. But there is also a race condition where restart happened in the middle of saving state. Any better suggestion, otherwise I will just add a TODO in the code and we can come back to that later.

File: agent/engine/docker_container_engine.go (hunk 8 above)
Comment (richardpen) [reply agent/engine/docker_container_engine.go]: removed

File: agent/engine/task_manager.go (hunk 9 above)
Comment (samuelkarp) [reply agent/engine/task_manager.go]: If the only downside is an error in the IPAM logs, I'm okay with adding a TODO.

**Comment by richardpen** [comment]
@samuelkarp I have addressed all your comments, merge this now. Feel free to continue comment on this if you have more comments.

File: agent/engine/engine_integ_test.go [hunk 11]
```diff
@@ -89,6 +89,7 @@ func setup(cfg *config.Config, t *testing.T) (TaskEngine, func(), credentials.Ma
 	}, credentialsManager
 }
 
+/*
```
Comment (samuelkarp) [inline agent/engine/engine_integ_test.go]: Why was this commented out?

File: agent/engine/docker_container_engine.go [hunk 12]
```diff
@@ -271,7 +274,7 @@ func (dg *dockerGoClient) pullImage(image string, authData *api.RegistryAuthenti
 
 	// Special case; this image is not one that should be pulled, but rather
 	// should be created locally if necessary
-	if image == emptyvolume.Image+":"+emptyvolume.Tag {
+	if image == fmt.Sprintf(imageNameFormat, emptyvolume.Image, emptyvolume.Tag) {
```
Comment (samuelkarp) [inline agent/engine/docker_container_engine.go]: Why was this changed?

File: agent/engine/engine_integ_test.go (hunk 11 above)
Comment (richardpen) [reply agent/engine/engine_integ_test.go]: This part should be removed when we merge the enis branch, as it is moved to `test_common.go`,
 see [here](https://github.com/aws/amazon-ecs-agent/blob/dev/agent/engine/test_common.go#L26), but l forgot to remove this part, I will remove this part later.

File: agent/engine/docker_container_engine.go (hunk 12 above)
Comment (richardpen) [reply agent/engine/docker_container_engine.go]: I thought this was better the previous way of doing this, and we will use this format for more internal image names.
